File: wisefy/src/androidTest/java/com/isupatches/wisefy/GetNearbyAccessPointsTests.java
package com.isupatches.wisefy;


import android.net.wifi.ScanResult;
import com.isupatches.wisefy.callbacks.GetNearbyAccessPointsCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import org.junit.Test;
import org.mockito.ArgumentMatchers;
import java.util.ArrayList;
import java.util.List;
import static com.isupatches.wisefy.base.TestUtils.TEST_SSID;
import static com.isupatches.wisefy.base.TestUtils.TEST_SSID2;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.fail;
import static org.mockito.Mockito.after;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class GetNearbyAccessPointsTests extends BaseAndroidJUnit4TestClass {

    @Test
    public void sync_failure_missingPrerequisite_filterDuplicates_false() {
        missingPrerequisite();
        assertEquals(null, mWiseFy.getNearbyAccessPoints(false));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
    }

    @Test
    public void sync_failure_missingPrerequisite_filterDuplicates_true() {
        missingPrerequisite();
        assertEquals(null, mWiseFy.getNearbyAccessPoints(true));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
    }

    @Test
    public void sync_success_filterDuplicates_false() {
        List<ScanResult> scanResults = new ArrayList<>();
        ScanResult scanResult1 = mock(ScanResult.class);
        scanResult1.SSID = TEST_SSID;
        scanResult1.level = -35;
        scanResults.add(scanResult1);
        ScanResult scanResult2 = mock(ScanResult.class);
        scanResult2.SSID = TEST_SSID2;
        scanResult2.level = -70;
        scanResults.add(scanResult2);

        when(mMockWiFiManager.getScanResults()).thenReturn(scanResults);

        List<ScanResult> accessPoints = mWiseFy.getNearbyAccessPoints(false);
        assertEquals(accessPoints, scanResults);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getScanResults();
    }

    @Test
    public void sync_success_filterDuplicates_true() {
        List<ScanResult> scanResults = new ArrayList<>();
        ScanResult scanResult1 = mock(ScanResult.class);
        scanResult1.SSID = TEST_SSID;
        scanResult1.level = -35;
        scanResults.add(scanResult1);
        ScanResult scanResult2 = mock(ScanResult.class);
        scanResult2.SSID = TEST_SSID2;
        scanResult2.level = -70;
        scanResults.add(scanResult2);

        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.removeEntriesWithLowerSignalStrength(ArgumentMatchers.<ScanResult>anyList())).thenReturn(scanResults);

        List<ScanResult> accessPoints = mWiseFy.getNearbyAccessPoints(true);
        assertEquals(accessPoints, scanResults);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getScanResults();
    }

    @Test
    public void async_failure_missingPrerequisite_filterDuplicates_false() {
        missingPrerequisite();
        GetNearbyAccessPointsCallbacks mockCallbacks = mock(GetNearbyAccessPointsCallbacks.class);
        mWiseFy.getNearbyAccessPoints(false, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getNearbyAccessPointsWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
    }

    @Test
    public void async_failure_missingPrerequisite_filterDuplicates_false_nullCallback() {
        missingPrerequisite();
        try {
            mWiseFy.getNearbyAccessPoints(false, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_missingPrerequisite_filterDuplicates_true() {
        missingPrerequisite();
        GetNearbyAccessPointsCallbacks mockCallbacks = mock(GetNearbyAccessPointsCallbacks.class);
        mWiseFy.getNearbyAccessPoints(true, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getNearbyAccessPointsWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
    }

    @Test
    public void async_failure_missingPrerequisite_filterDuplicates_true_nullCallback() {
        missingPrerequisite();
        try {
            mWiseFy.getNearbyAccessPoints(true, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success_filterDuplicates_false() {
        List<ScanResult> scanResults = new ArrayList<>();
        ScanResult scanResult1 = mock(ScanResult.class);
        scanResult1.SSID = TEST_SSID;
        scanResult1.level = -35;
        scanResults.add(scanResult1);
        ScanResult scanResult2 = mock(ScanResult.class);
        scanResult2.SSID = TEST_SSID2;
        scanResult2.level = -70;
        scanResults.add(scanResult2);

        when(mMockWiFiManager.getScanResults()).thenReturn(scanResults);

        GetNearbyAccessPointsCallbacks mockCallbacks = mock(GetNearbyAccessPointsCallbacks.class);
        mWiseFy.getNearbyAccessPoints(false, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedNearbyAccessPoints(scanResults);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getScanResults();
    }

    @Test
    public void async_success_filterDuplicates_false_nullCallback() {
        List<ScanResult> scanResults = new ArrayList<>();
        ScanResult scanResult1 = mock(ScanResult.class);
        scanResult1.SSID = TEST_SSID;
        scanResult1.level = -35;
        scanResults.add(scanResult1);
        ScanResult scanResult2 = mock(ScanResult.class);
        scanResult2.SSID = TEST_SSID2;
        scanResult2.level = -70;
        scanResults.add(scanResult2);

        when(mMockWiFiManager.getScanResults()).thenReturn(scanResults);

        try {
            mWiseFy.getNearbyAccessPoints(false, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success_filterDuplicates_true() {
        List<ScanResult> scanResults = new ArrayList<>();
        ScanResult scanResult1 = mock(ScanResult.class);
        scanResult1.SSID = TEST_SSID;
        scanResult1.level = -35;
        scanResults.add(scanResult1);
        ScanResult scanResult2 = mock(ScanResult.class);
        scanResult2.SSID = TEST_SSID2;
        scanResult2.level = -70;
        scanResults.add(scanResult2);

        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.removeEntriesWithLowerSignalStrength(ArgumentMatchers.<ScanResult>anyList())).thenReturn(scanResults);

        GetNearbyAccessPointsCallbacks mockCallbacks = mock(GetNearbyAccessPointsCallbacks.class);
        mWiseFy.getNearbyAccessPoints(true, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedNearbyAccessPoints(scanResults);
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).getScanResults();
    }

    @Test
    public void async_success_filterDuplicates_true_nullCallback() {
        List<ScanResult> scanResults = new ArrayList<>();
        ScanResult scanResult1 = mock(ScanResult.class);
        scanResult1.SSID = TEST_SSID;
        scanResult1.level = -35;
        scanResults.add(scanResult1);
        ScanResult scanResult2 = mock(ScanResult.class);
        scanResult2.SSID = TEST_SSID2;
        scanResult2.level = -70;
        scanResults.add(scanResult2);

        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.removeEntriesWithLowerSignalStrength(ArgumentMatchers.<ScanResult>anyList())).thenReturn(scanResults);

        try {
            mWiseFy.getNearbyAccessPoints(true, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).getScanResults();
        } catch (NullPointerException npe) {
            fail();
        }
    }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/GetNearbyAccessPointsTests.kt
package com.isupatches.wisefy.main

import com.isupatches.wisefy.BaseAndroidJUnit4TestClass
import com.isupatches.wisefy.VERIFICATION_SUCCESS_TIMEOUT
import com.isupatches.wisefy.callbacks.GetNearbyAccessPointsCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER

import org.junit.Assert.assertEquals
import org.junit.Test

import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to retrieve a list of nearby access points for a device.
 *
 * @author Patches
 */
internal class GetNearbyAccessPointsTests : BaseAndroidJUnit4TestClass() {

    @Test fun sync_failure_prechecks_filterDuplicates_false() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_failure()
        assertEquals(null, wisefy.getNearbyAccessPoints(false))
    }

    @Test fun sync_failure_prechecks_filterDuplicates_true() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_failure()
        assertEquals(null, wisefy.getNearbyAccessPoints(true))
    }

    @Test fun sync_failure_filterDuplicates_false() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_null(false)
        val nearbyAccessPoints = wisefy.getNearbyAccessPoints(false)
        assertEquals(null, nearbyAccessPoints)
    }

    @Test fun sync_failure_filterDuplicates_true() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_null(true)
        val nearbyAccessPoints = wisefy.getNearbyAccessPoints(true)
        assertEquals(null, nearbyAccessPoints)
    }

    @Test fun sync_success_filterDuplicates_false() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        val accessPoints = mockWiseFySearchUtil.nearbyAccessPoints_success(false)
        val nearbyAccessPoints = wisefy.getNearbyAccessPoints(false)
        assertEquals(accessPoints, nearbyAccessPoints)
    }

    @Test fun sync_success_filterDuplicates_true() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        val accessPoints = mockWiseFySearchUtil.nearbyAccessPoints_success(true)
        val nearbyAccessPoints = wisefy.getNearbyAccessPoints(true)
        assertEquals(accessPoints, nearbyAccessPoints)
    }

    @Test fun async_failure_prechecks_filterDuplicates_false() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_failure()
        val mockCallbacks = mock(GetNearbyAccessPointsCallbacks::class.java)
        wisefy.getNearbyAccessPoints(false, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
    }

    @Test fun async_failure_prechecks_filterDuplicates_false_nullCallback() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_failure()
        nullCallbackUtil.callGetNearbyAccessPoints(false)
    }

    @Test fun async_failure_prechecks_filterDuplicates_true() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_failure()
        val mockCallbacks = mock(GetNearbyAccessPointsCallbacks::class.java)
        wisefy.getNearbyAccessPoints(true, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
    }

    @Test fun async_failure_prechecks_filterDuplicates_true_nullCallback() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_failure()
        nullCallbackUtil.callGetNearbyAccessPoints(true)
    }

    @Test fun async_failure_filterDuplicates_false() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_null(false)
        val mockCallbacks = mock(GetNearbyAccessPointsCallbacks::class.java)
        wisefy.getNearbyAccessPoints(false, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noAccessPointsFound()
    }

    @Test fun async_failure_filterDuplicates_false_nullCallback() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_success(true)
        nullCallbackUtil.callGetNearbyAccessPoints(false)
    }

    @Test fun async_failure_filterDuplicates_true() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_null(true)
        val mockCallbacks = mock(GetNearbyAccessPointsCallbacks::class.java)
        wisefy.getNearbyAccessPoints(true, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).noAccessPointsFound()
    }

    @Test fun async_failure_filterDuplicates_true_nullCallback() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_null(true)
        nullCallbackUtil.callGetNearbyAccessPoints(true)
    }

    @Test fun async_success_filterDuplicates_false() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        val accessPoints = mockWiseFySearchUtil.nearbyAccessPoints_success(false)
        val mockCallbacks = mock(GetNearbyAccessPointsCallbacks::class.java)
        wisefy.getNearbyAccessPoints(false, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedNearbyAccessPoints(accessPoints)
    }

    @Test fun async_success_filterDuplicates_false_nullCallback() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_success(true)
        nullCallbackUtil.callGetNearbyAccessPoints(false)
    }

    @Test fun async_success_filterDuplicates_true() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        val accessPoints = mockWiseFySearchUtil.nearbyAccessPoints_success(true)
        val mockCallbacks = mock(GetNearbyAccessPointsCallbacks::class.java)
        wisefy.getNearbyAccessPoints(true, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedNearbyAccessPoints(accessPoints)
    }

    @Test fun async_success_filterDuplicates_true_nullCallback() {
        mockWiseFyPrechecksUtil.getNearbyAccessPoints_success()
        mockWiseFySearchUtil.nearbyAccessPoints_success(true)
        nullCallbackUtil.callGetNearbyAccessPoints(true)
    }
}
